### ISSUE TITLE
Scoop: 4095 The parameter is incorrect

### ISSUE THREAD

**Issue by Have-a-question-about-this-project** [JohnnyFFM/Poc1to2Converter#16]
Running v.2.3
Disk to disk conversion of file XXXXXXXXXXXX_0_2661664_2661664
  I/O Error -Seek to read
  Scoop: 4095 The parameter is incorrect
This happens as soon as it starts Processing Scoop Pairs 1/2048

**Comment by JohnnyFFM** [comment]
Hi!

The good news first: your file is untouched, no damage (always the case if disk-to-disk). 

To further investigate, I would need more info: did you check the fast disk-to-disk option? And are you on usb? Do you use raid? 

The error means that the app failed to seek to scoop 4095 (the very last scoop). This looks to me like an addressing issue when using direct i/o (i/o controller issue / exotic disk format). 

Two things that might solve your problem:
First, in case it‘s an usb drive, try to connect it directly to your pc, no hub in between. 

Second, disable direct I/O. This can be done by starting the converter with the command line option -ddio.

Before you retry, delete the empty file on the target drive. 


I‘ll run a couple of tests on my side as well, but will not be able to do so before weekend.

**Comment by Have-a-question-about-this-project** [comment]
This was on a LaCie 8TB USB3 disk. It might be an issue with 4K blocks if you're not catering for that.
Its directly connected, no hub.
Its actually the first file I tried directly on the disk as this file doesn't fit on my SSD,
I haven't ticked "Fast Disk to Disk mode" or "Speed Boost" as they were both marked experimental and I'm not clear on what they do.
I'll try disabling direct IO after you confirm that there isn't an issue with 4K drives.

I've not been using it directly as I the disk is SMR (shingled) so it performs badly writing big files if its not fed a sequential stream. Originally I plotted poc1 files with a large memory buffer resulting in writes of 2.5M, which to the drive was as good as random as it was 4096 separate streams of 2.5M writes. It ground to a complete halt after a few minutes as it had too much data to relocate. For this reason I scripted it to plot to SSD then copy to SDD after.

Does your converter write in one clean sequential stream when converting disk to disk? Or even for same disk conversion? I've typically got 400G plots. It'd be nice if you had an option to buffer to SSD to allow for less disk seeking. Then It could do something like read 100G, write 100G rather than lots of smaller bursts.

I've also got 32GB RAM, so your max of 4096M seems a little low to me.

Ultimately I've got 7x 8TB USB drives left to convert, so it'd be nice if they could all convert in parallel running one instance of your tool while sharing large memory and SSD buffers.

I should probably have split this off into several bugs/feature requests.

Other ideas - stop disk IO during a burst block for a configurable amount of seconds so there isn't a battle for the disk head if the miner isn't stopped from mining the disk being converted.
Put the file being converted into a subdir or rename it to stop the miner from seeing it while its being converted.
Allow drag and drop of files/dirs to be converted.

**Comment by JohnnyFFM** [comment]
Hi! Thanks for all the details.
Let‘s start with the 4k block. My tool is checking logical block size and checks if direct i/o would work. But there might be some rare cases where it fails (some usb bridges dont handle it properly). My general advice: if you have a block size of 4k always plot a multiple of 64 nonces to align nonces and sectors. I do recommend to test it with the flag -ddio (as you use disk-to-disk there won‘t be any harm if it fails). It will use a stream that can handle adressing inside sectors.

It‘s safe to use fast disk-to-disk, i‘ll remove the experimental label next version. Its a bit faster (~30%)

Regarding SMR: conversion is also fast on smr. 8tb take 16h disk-to-disk and 32-48h inplace. You can increase the speed a little with the boost option (uses more ram).
I was testing this with 400gb files as well.
No ssd cache needed!

A lot of ppl are stating that 4gb is too low, however, there is no real speed improvement if you use more. Conversion is i/o intense and doesnt need cpu. Ram is only nice to have for smr, but 4gb is really more than needed for current drive capacities.

You can convert in parallel, but on usb i would not advise to convert more than 2 drives per controller at the same time (for inplace conversion, for disk-to-disk only 1 conversion at the same time).

Dont mine and convert a drive at the same time. A conversion will stress drive i/o to the max. An additional read would slow things down. Since conversion is a one time thing, there is no added value coding a difficult pause-conversion-for-mining feature. 

Drag&Drop is a very nice idea 👌

Hope this helps. Let me know if the -ddio does the trick. Any further question please let me know.

Best,

Johnny

**Comment by JohnnyFFM** [comment]
Hi, any news on this?

**Comment by Have-a-question-about-this-project** [comment]
Hi. Before I got you last message about the ddio flag I did try the other files on the drive and it didn't like them either - they were 400GB. I then tried ticking fast disk to disk (not sure what you're doing to boost it) and the speed boot option and it was happy to do same disk conversion. This completed OK. I have one other LaCie disk and that completed 4 files then broke on the 5th with "ERROR! Error writing back scoop : 64". I replotted that one file and got it to continue and it did the other 15 or so files fine. Unfortunately I've been busy and didn't have the chance to try the ddio option. I might have if I had a way to to stop the replot without killing the current file. You don't have a way to pause/resume or stop after the current file, the only way to stop it is to kill the program and destroy the file currently being converted.
I've tried my Seagarte 8TB Hub drives any they don't have the same issue the LaCie drives did.
What does the fast disk to disk option do differently?

**Comment by JohnnyFFM** [comment]
Hi, 
good to hear to you've managed to make it work.

The fast disk-to-disk reads and writes in parallel rather then just alternating reads and writes. 

The speed boost option uses more ram. I guess this option made it work for you .
Background: assuming your drive/usb Bridge of the LaCies uses 4k native sector format, but reports like a 512e sector format. My tool will not disable direct i/o in this case. As your number of nonces is not a multiple of 64 but 32 the converter performs a seek to the middle of a sector which fails in direct i/o. Since the boost option is reading more scoops at once, i.e. adressing, reading and writing is multiplied with a factor of  2^x, it fixes the issue. 
Alternative: the sector size check I perform to assess direct i/o compatibilty  has a bug, I'll have a look into this as well. 

The "ERROR! Error writing back scoop : >0"  means the write failed most likely due to a harddisk issue. If replotting that file made it work, all good.

Pause / Resume: You can make the converter stop after the current file in the queue. Just press the top-right X to close the converter. It will ask you to either abort (yes), to stop after current file in queue (no) or to cancel and just continue your conversion. 

Rgds,

Johnny
